dot/digest/digest.go

```diff
@@ -91,44 +76,80 @@ func (h *Handler) Stop() error {
 	return nil
 }
 
-// NextGrandpaAuthorityChange returns the block number of the next upcoming grandpa authorities change.
-// It returns 0 if no change is scheduled.
-func (h *Handler) NextGrandpaAuthorityChange() (next uint) {
-	next = ^uint(0)
-
-	if h.grandpaScheduledChange != nil {
-		next = h.grandpaScheduledChange.atBlock
+// HandleDigests handles consensus digests for an imported block
+func (h *Handler) HandleDigests(header *types.Header) error {
+	consensusDigests := h.toConsensusDigests(header.Digest.Types)
+	consensusDigests, err := checkForGRANDPAForcedChanges(consensusDigests)
+	if err != nil {
+		return fmt.Errorf("failed while checking GRANDPA digests: %w", err)
 	}
 
-	if h.grandpaForcedChange != nil && h.grandpaForcedChange.atBlock < next {
-		next = h.grandpaForcedChange.atBlock
+	for i := range consensusDigests {
+		// avoiding implicit memory aliasing in for loop, since:
+		// for _, digest := range consensusDigests { &digest }
+		// is using the address of a loop variable
+		digest := consensusDigests[i]
+		err := h.handleConsensusDigest(&digest, header)
+		if err != nil {
+			h.logger.Errorf("cannot handle consensus digest: %w", err)
+		}
 	}
 
-	if h.grandpaPause != nil && h.grandpaPause.atBlock < next {
-		next = h.grandpaPause.atBlock
-	}
+	return nil
+}
+
+// toConsensusDigests converts a slice of scale.VaryingDataType to a slice of types.ConsensusDigest.
+func (h *Handler) toConsensusDigests(scaleVaryingTypes []scale.VaryingDataType) []types.ConsensusDigest {
+	consensusDigests := make([]types.ConsensusDigest, 0, len(scaleVaryingTypes))
+
+	for _, d := range scaleVaryingTypes {
+		digest, ok := d.Value().(types.ConsensusDigest)
+		if !ok {
+			h.logger.Debugf("digest type not supported: %T", d.Value())
+			continue
+		}
 
-	if h.grandpaResume != nil && h.grandpaResume.atBlock < next {
-		next = h.grandpaResume.atBlock
+		switch digest.ConsensusEngineID {
+		case types.GrandpaEngineID, types.BabeEngineID:
+			consensusDigests = append(consensusDigests, digest)
+		}
 	}
 
-	return next
+	return consensusDigests
 }
 
-// HandleDigests handles consensus digests for an imported block
-func (h *Handler) HandleDigests(header *types.Header) {
-	for i, d := range header.Digest.Types {
-		val, ok := d.Value().(types.ConsensusDigest)
-		if !ok {
+// checkForGRANDPAForcedChanges removes any GrandpaScheduledChange in the presence of a
+// GrandpaForcedChange in the same block digest, returning a new slice of types.ConsensusDigest
+func checkForGRANDPAForcedChanges(digests []types.ConsensusDigest) ([]types.ConsensusDigest, error) {
+	var hasForcedChange bool
+	digestsWithoutScheduled := make([]types.ConsensusDigest, 0, len(digests))
```

Comment:
why are defining the capacity of the slice, but using `append` on lines 142 and 144.  `append` will actually double the capacity if it overflows.

Reply:
Yes, but in this case, the maximum amount of data in the `digestsWithoutScheduled` slice is `len(digest)`, so there is no problem to use append as it will never overflow the capacity